File: src/lib/db/repositories/buildRepository.ts
import { and, eq, inArray } from "drizzle-orm";

import type { ConceptTagId } from "@/data/conceptTags";
import type { AppDatabase } from "@/lib/db/client";
import { buildSynergyTypes, buildTags, builds, buildVariants } from "@/lib/db/schema";
import type { SynergyTypeDesignation } from "@/lib/builds/resolveDesignatedSynergies";
import {
  parseSoftStatTargets,
  serializeSoftStatTargets,
  type SoftStatTargets,
} from "@/lib/builds/softStatTargets";
import type { SynergyType } from "@/lib/synergies/schemas";

export type BuildRecord = {
  id: string;
  userId: number;
  name: string;
  className: string;
  subclass: unknown;
  exoticArmorHash: number | null;
  exoticArmorName: string | null;
  exoticWeaponHash: number | null;
  exoticWeaponName: string | null;
  pinnedSuper: string | null;
  softStatTargets: SoftStatTargets;
  tagIds: ConceptTagId[];
  synergyTypes: SynergyTypeDesignation[];
  createdAt: string;
  updatedAt: string;
};

function mapBuildSynergyTypeRow(r: {
  type: string;
  subType: string | null;
}): SynergyTypeDesignation {
  return {
    type: r.type as SynergyType,
    // Empty string stored for UNIQUE null-safety; expose as null
    subType: r.subType?.trim() ? r.subType : null,
  };
}

function loadBuildTags(db: AppDatabase, buildId: string): ConceptTagId[] {
  return db
    .select({ tagId: buildTags.tagId })
    .from(buildTags)
    .where(eq(buildTags.buildId, buildId))
    .all()
    .map((r) => r.tagId as ConceptTagId)
    .sort();
}

function loadBuildSynergyTypes(db: AppDatabase, buildId: string): SynergyTypeDesignation[] {
  return db
    .select({
      type: buildSynergyTypes.type,
      subType: buildSynergyTypes.subType,
    })
    .from(buildSynergyTypes)
    .where(eq(buildSynergyTypes.buildId, buildId))
    .all()
    .map(mapBuildSynergyTypeRow);
}

/** Batch-load tags for many builds (one query). Tag ids sorted per build. */
function loadBuildTagsForIds(db: AppDatabase, buildIds: string[]): Map<string, ConceptTagId[]> {
  const map = new Map<string, ConceptTagId[]>();
  for (const id of buildIds) map.set(id, []);
  if (buildIds.length === 0) return map;

  const rows = db
    .select({ buildId: buildTags.buildId, tagId: buildTags.tagId })
    .from(buildTags)
    .where(inArray(buildTags.buildId, buildIds))
    .all();

  for (const row of rows) {
    const list = map.get(row.buildId);
    if (list) list.push(row.tagId as ConceptTagId);
    else map.set(row.buildId, [row.tagId as ConceptTagId]);
  }
  for (const [id, tags] of map) {
    map.set(id, tags.sort());
  }
  return map;
}

/** Batch-load synergy type designations for many builds (one query). */
function loadBuildSynergyTypesForIds(
  db: AppDatabase,
  buildIds: string[],
): Map<string, SynergyTypeDesignation[]> {
  const map = new Map<string, SynergyTypeDesignation[]>();
  for (const id of buildIds) map.set(id, []);
  if (buildIds.length === 0) return map;

  const rows = db
    .select({
      buildId: buildSynergyTypes.buildId,
      type: buildSynergyTypes.type,
      subType: buildSynergyTypes.subType,
    })
    .from(buildSynergyTypes)
    .where(inArray(buildSynergyTypes.buildId, buildIds))
    .all();

  for (const row of rows) {
    const designation = mapBuildSynergyTypeRow(row);
    const list = map.get(row.buildId);
    if (list) list.push(designation);
    else map.set(row.buildId, [designation]);
  }
  return map;
}

function rowToBuildCore(
  row: typeof builds.$inferSelect,
  tagIds: ConceptTagId[],
  synergyTypes: SynergyTypeDesignation[],
): BuildRecord {
  return {
    id: row.id,
    userId: row.userId,
    name: row.name,
    className: row.className,
    subclass: JSON.parse(row.subclass) as unknown,
    exoticArmorHash: row.exoticArmorHash ?? null,
    exoticArmorName: row.exoticArmorName ?? null,
    exoticWeaponHash: row.exoticWeaponHash ?? null,
    exoticWeaponName: row.exoticWeaponName ?? null,
    pinnedSuper: row.pinnedSuper ?? null,
    softStatTargets: parseSoftStatTargets(row.softStatTargets),
    tagIds,
    synergyTypes,
    createdAt: row.createdAt,
    updatedAt: row.updatedAt,
  };
}

function rowToBuild(db: AppDatabase, row: typeof builds.$inferSelect): BuildRecord {
  return rowToBuildCore(row, loadBuildTags(db, row.id), loadBuildSynergyTypes(db, row.id));
}

/** Hydrate many build rows with two child queries total. */
function rowsToBuilds(db: AppDatabase, rows: (typeof builds.$inferSelect)[]): BuildRecord[] {
  if (rows.length === 0) return [];
  const ids = rows.map((r) => r.id);
  const tagsByBuild = loadBuildTagsForIds(db, ids);
  const synergyByBuild = loadBuildSynergyTypesForIds(db, ids);
  return rows.map((row) =>
    rowToBuildCore(row, tagsByBuild.get(row.id) ?? [], synergyByBuild.get(row.id) ?? []),
  );
}

export function listBuilds(db: AppDatabase, userId: number): BuildRecord[] {
  const rows = db.select().from(builds).where(eq(builds.userId, userId)).all();
  return rowsToBuilds(db, rows);
}

export function listBuildsByTags(
  db: AppDatabase,
  userId: number,
  tagIds: ConceptTagId[],
): BuildRecord[] {
  if (tagIds.length === 0) return listBuilds(db, userId);

  let matching: string[] | null = null;
  for (const tagId of tagIds) {
    const ids = db
      .select({ buildId: buildTags.buildId })
      .from(buildTags)
      .innerJoin(builds, eq(buildTags.buildId, builds.id))
      .where(and(eq(buildTags.tagId, tagId), eq(builds.userId, userId)))
      .all()
      .map((r) => r.buildId);
    matching = matching === null ? ids : matching.filter((id) => ids.includes(id));
  }
  if (!matching || matching.length === 0) return [];

  const rows = db
    .select()
    .from(builds)
    .where(and(eq(builds.userId, userId), inArray(builds.id, matching)))
    .all();
  return rowsToBuilds(db, rows);
}

export function listBuildsFiltered(
  db: AppDatabase,
  userId: number,
  filters: {
    tags?: ConceptTagId[];
    exoticArmorHash?: number;
    exoticWeaponHash?: number;
    synergyType?: string;
    synergySubType?: string | null;
  },
): BuildRecord[] {
  let results = filters.tags?.length
    ? listBuildsByTags(db, userId, filters.tags)
    : listBuilds(db, userId);

  if (filters.exoticArmorHash !== undefined) {
    results = results.filter((b) => b.exoticArmorHash === filters.exoticArmorHash);
  }

  if (filters.synergyType) {
    const type = filters.synergyType;
    const sub =
      filters.synergySubType === undefined
        ? undefined
        : filters.synergySubType?.trim() || null;
    results = results.filter((b) =>
      b.synergyTypes.some((d) => {
        if (d.type !== type) return false;
        if (sub === undefined) return true;
        return (d.subType ?? null) === sub;
      }),
    );
  }

  if (filters.exoticWeaponHash !== undefined) {
    const weaponHash = filters.exoticWeaponHash;
    const buildIds = results.map((b) => b.id);
    if (buildIds.length === 0) return [];
    const variantRows = db
      .select({ buildId: buildVariants.buildId })
      .from(buildVariants)
      .where(
        and(
          inArray(buildVariants.buildId, buildIds),
          eq(buildVariants.exoticWeaponHash, weaponHash),
        ),
      )
      .all();
    const allowed = new Set(variantRows.map((r) => r.buildId));
    results = results.filter(
      (b) => b.exoticWeaponHash === weaponHash || allowed.has(b.id),
    );
  }

  return results;
}

export function findBuildByNameClass(
  db: AppDatabase,
  userId: number,
  className: string,
  name: string,
  excludeId?: string,
): BuildRecord | null {
  const rows = db
    .select()
    .from(builds)
    .where(and(eq(builds.userId, userId), eq(builds.className, className), eq(builds.name, name)))
    .all()
    .map((row) => rowToBuild(db, row));
  return rows.find((b) => b.id !== excludeId) ?? null;
}

export function getBuild(db: AppDatabase, userId: number, id: string): BuildRecord | null {
  const row = db
    .select()
    .from(builds)
    .where(and(eq(builds.id, id), eq(builds.userId, userId)))
    .get();
  return row ? rowToBuild(db, row) : null;
}

function insertSynergyTypes(
  db: AppDatabase,
  buildId: string,
  designations: SynergyTypeDesignation[],
  now: string,
): void {
  for (const d of designations) {
    db.insert(buildSynergyTypes)
      .values({
        buildId,
        type: d.type,
        // SQLite UNIQUE treats NULLs as distinct — use "" for no subType
        subType: d.subType?.trim() || "",
        attachedAt: now,
      })
      .run();
  }
}

export function createBuildRecord(
  db: AppDatabase,
  userId: number,
  input: {
    id: string;
    name: string;
    className: string;
    subclass: unknown;
    exoticArmorHash: number | null;
    exoticArmorName: string | null;
    exoticWeaponHash: number | null;
    exoticWeaponName: string | null;
    pinnedSuper: string | null;
    softStatTargets?: SoftStatTargets;
    tagIds: ConceptTagId[];
    synergyTypes: SynergyTypeDesignation[];
    now: string;
  },
): BuildRecord {
  db.insert(builds)
    .values({
      id: input.id,
      userId,
      name: input.name,
      className: input.className,
      subclass: JSON.stringify(input.subclass),
      exoticArmorHash: input.exoticArmorHash,
      exoticArmorName: input.exoticArmorName,
      exoticWeaponHash: input.exoticWeaponHash,
      exoticWeaponName: input.exoticWeaponName,
      pinnedSuper: input.pinnedSuper,
      softStatTargets: serializeSoftStatTargets(input.softStatTargets ?? {}),
      createdAt: input.now,
      updatedAt: input.now,
    })
    .run();

  for (const tagId of input.tagIds) {
    db.insert(buildTags).values({ buildId: input.id, tagId }).run();
  }
  insertSynergyTypes(db, input.id, input.synergyTypes, input.now);

  return getBuild(db, userId, input.id)!;
}

export function updateBuildRecord(
  db: AppDatabase,
  userId: number,
  id: string,
  patch: {
    name?: string;
    className?: string;
    subclass?: unknown;
    exoticArmorHash?: number | null;
    exoticArmorName?: string | null;
    exoticWeaponHash?: number | null;
    exoticWeaponName?: string | null;
    pinnedSuper?: string | null;
    softStatTargets?: SoftStatTargets;
    tagIds?: ConceptTagId[];
    synergyTypes?: SynergyTypeDesignation[];
    now: string;
  },
): BuildRecord | null {
  const existing = getBuild(db, userId, id);
  if (!existing) return null;

  db.update(builds)
    .set({
      name: patch.name ?? existing.name,
      className: patch.className ?? existing.className,
      subclass: patch.subclass ? JSON.stringify(patch.subclass) : JSON.stringify(existing.subclass),
      exoticArmorHash:
        patch.exoticArmorHash !== undefined ? patch.exoticArmorHash : existing.exoticArmorHash,
      exoticArmorName:
        patch.exoticArmorName !== undefined ? patch.exoticArmorName : existing.exoticArmorName,
      exoticWeaponHash:
        patch.exoticWeaponHash !== undefined ? patch.exoticWeaponHash : existing.exoticWeaponHash,
      exoticWeaponName:
        patch.exoticWeaponName !== undefined ? patch.exoticWeaponName : existing.exoticWeaponName,
      pinnedSuper: patch.pinnedSuper !== undefined ? patch.pinnedSuper : existing.pinnedSuper,
      softStatTargets:
        patch.softStatTargets !== undefined
          ? serializeSoftStatTargets(patch.softStatTargets)
          : serializeSoftStatTargets(existing.softStatTargets),
      updatedAt: patch.now,
    })
    .where(and(eq(builds.id, id), eq(builds.userId, userId)))
    .run();

  if (patch.tagIds) {
    db.delete(buildTags).where(eq(buildTags.buildId, id)).run();
    for (const tagId of patch.tagIds) {
      db.insert(buildTags).values({ buildId: id, tagId }).run();
    }
  }

  if (patch.synergyTypes) {
    db.delete(buildSynergyTypes).where(eq(buildSynergyTypes.buildId, id)).run();
    insertSynergyTypes(db, id, patch.synergyTypes, patch.now);
  }

  return getBuild(db, userId, id);
}

export function deleteBuildRecord(db: AppDatabase, userId: number, id: string): boolean {
  const result = db
    .delete(builds)
    .where(and(eq(builds.id, id), eq(builds.userId, userId)))
    .run();
  return result.changes > 0;
}

File: flutter/packages/db/lib/src/repos/build_repository.dart
import 'package:drift/drift.dart';

import '../app_database.dart';
import '../tables.dart';
import 'json_codec.dart';
import 'library_records.dart';

BuildRecord _rowToBuild(
  Build row,
  List<String> tagIds,
  List<SynergyTypeDesignationRecord> synergyTypes,
) {
  return BuildRecord(
    id: row.id,
    userId: row.userId,
    name: row.name,
    className: row.className,
    subclass: decodeJsonValue(row.subclass),
    exoticArmorHash: row.exoticArmorHash,
    exoticArmorName: row.exoticArmorName,
    exoticWeaponHash: row.exoticWeaponHash,
    exoticWeaponName: row.exoticWeaponName,
    pinnedSuper: row.pinnedSuper,
    softStatTargets: parseSoftStatTargetsJson(row.softStatTargets),
    tagIds: tagIds,
    synergyTypes: synergyTypes,
    createdAt: row.createdAt,
    updatedAt: row.updatedAt,
  );
}

Future<List<String>> _loadBuildTags(AppDatabase db, String buildId) async {
  final rows = await (db.select(db.buildTags)
        ..where((t) => t.buildId.equals(buildId)))
      .get();
  final tags = rows.map((r) => r.tagId).toList()..sort();
  return tags;
}

Future<List<SynergyTypeDesignationRecord>> _loadBuildSynergyTypes(
  AppDatabase db,
  String buildId,
) async {
  final rows = await (db.select(db.buildSynergyTypes)
        ..where((t) => t.buildId.equals(buildId)))
      .get();
  return rows
      .map(
        (r) => SynergyTypeDesignationRecord(
          type: r.type,
          subType: (r.subType?.trim().isEmpty ?? true) ? null : r.subType,
        ),
      )
      .toList();
}

Future<Map<String, List<String>>> _loadBuildTagsForIds(
  AppDatabase db,
  List<String> buildIds,
) async {
  final map = {for (final id in buildIds) id: <String>[]};
  if (buildIds.isEmpty) return map;
  final rows = await (db.select(db.buildTags)
        ..where((t) => t.buildId.isIn(buildIds)))
      .get();
  for (final row in rows) {
    map.putIfAbsent(row.buildId, () => <String>[]).add(row.tagId);
  }
  for (final id in map.keys) {
    map[id]!.sort();
  }
  return map;
}

Future<Map<String, List<SynergyTypeDesignationRecord>>>
    _loadBuildSynergyTypesForIds(
  AppDatabase db,
  List<String> buildIds,
) async {
  final map = {
    for (final id in buildIds) id: <SynergyTypeDesignationRecord>[],
  };
  if (buildIds.isEmpty) return map;
  final rows = await (db.select(db.buildSynergyTypes)
        ..where((t) => t.buildId.isIn(buildIds)))
      .get();
  for (final row in rows) {
    final d = SynergyTypeDesignationRecord(
      type: row.type,
      subType: (row.subType?.trim().isEmpty ?? true) ? null : row.subType,
    );
    map.putIfAbsent(row.buildId, () => <SynergyTypeDesignationRecord>[]).add(d);
  }
  return map;
}

Future<List<BuildRecord>> _rowsToBuilds(
  AppDatabase db,
  List<Build> rows,
) async {
  if (rows.isEmpty) return const [];
  final ids = rows.map((r) => r.id).toList();
  final tagsByBuild = await _loadBuildTagsForIds(db, ids);
  final synergyByBuild = await _loadBuildSynergyTypesForIds(db, ids);
  return rows
      .map(
        (row) => _rowToBuild(
          row,
          tagsByBuild[row.id] ?? const [],
          synergyByBuild[row.id] ?? const [],
        ),
      )
      .toList();
}

Future<void> _insertSynergyTypes(
  AppDatabase db,
  String buildId,
  List<SynergyTypeDesignationRecord> designations,
  String now,
) async {
  for (final d in designations) {
    final sub =
        d.subType != null && d.subType!.trim().isNotEmpty ? d.subType! : '';
    await db.into(db.buildSynergyTypes).insert(
          BuildSynergyTypesCompanion.insert(
            buildId: buildId,
            type: d.type,
            subType: Value(sub),
            attachedAt: now,
          ),
        );
  }
}

/// List all builds for [userId].
Future<List<BuildRecord>> listBuilds(AppDatabase db, int userId) async {
  final rows = await (db.select(db.builds)
        ..where((t) => t.userId.equals(userId)))
      .get();
  return _rowsToBuilds(db, rows);
}

/// Get one build scoped to [userId], or null.
Future<BuildRecord?> getBuild(
  AppDatabase db,
  int userId,
  String id,
) async {
  final row = await (db.select(db.builds)
        ..where((t) => t.id.equals(id) & t.userId.equals(userId)))
      .getSingleOrNull();
  if (row == null) return null;
  final tags = await _loadBuildTags(db, row.id);
  final types = await _loadBuildSynergyTypes(db, row.id);
  return _rowToBuild(row, tags, types);
}

/// Create a build with optional tags and synergy type designations.
Future<BuildRecord> createBuildRecord(
  AppDatabase db,
  int userId, {
  required String id,
  required String name,
  required String className,
  Object? subclass,
  int? exoticArmorHash,
  String? exoticArmorName,
  int? exoticWeaponHash,
  String? exoticWeaponName,
  String? pinnedSuper,
  Map<String, Object?>? softStatTargets,
  List<String> tagIds = const [],
  List<SynergyTypeDesignationRecord> synergyTypes = const [],
  required String now,
}) async {
  await db.into(db.builds).insert(
        BuildsCompanion.insert(
          id: id,
          userId: userId,
          name: name,
          className: className,
          subclass: encodeJsonValue(subclass ?? const <String, Object?>{}),
          exoticArmorHash: Value(exoticArmorHash),
          exoticArmorName: Value(exoticArmorName),
          exoticWeaponHash: Value(exoticWeaponHash),
          exoticWeaponName: Value(exoticWeaponName),
          pinnedSuper: Value(pinnedSuper),
          softStatTargets: Value(encodeSoftStatTargetsJson(softStatTargets)),
          createdAt: now,
          updatedAt: now,
        ),
      );
  for (final tagId in tagIds) {
    await db.into(db.buildTags).insert(
          BuildTagsCompanion.insert(buildId: id, tagId: tagId),
        );
  }
  await _insertSynergyTypes(db, id, synergyTypes, now);
  return (await getBuild(db, userId, id))!;
}

/// Patch a build; returns null if missing.
///
/// Nullable fields use [Value] wrappers: omit (null param) keeps existing;
/// pass [Value] with null to clear.
Future<BuildRecord?> updateBuildRecord(
  AppDatabase db,
  int userId,
  String id, {
  String? name,
  String? className,
  Object? subclass,
  Value<int?> exoticArmorHash = const Value.absent(),
  Value<String?> exoticArmorName = const Value.absent(),
  Value<int?> exoticWeaponHash = const Value.absent(),
  Value<String?> exoticWeaponName = const Value.absent(),
  Value<String?> pinnedSuper = const Value.absent(),
  Map<String, Object?>? softStatTargets,
  List<String>? tagIds,
  List<SynergyTypeDesignationRecord>? synergyTypes,
  required String now,
}) async {
  final existing = await getBuild(db, userId, id);
  if (existing == null) return null;

  await (db.update(db.builds)
        ..where((t) => t.id.equals(id) & t.userId.equals(userId)))
      .write(
    BuildsCompanion(
      name: Value(name ?? existing.name),
      className: Value(className ?? existing.className),
      subclass: Value(encodeJsonValue(subclass ?? existing.subclass)),
      exoticArmorHash: exoticArmorHash.present
          ? exoticArmorHash
          : Value(existing.exoticArmorHash),
      exoticArmorName: exoticArmorName.present
          ? exoticArmorName
          : Value(existing.exoticArmorName),
      exoticWeaponHash: exoticWeaponHash.present
          ? exoticWeaponHash
          : Value(existing.exoticWeaponHash),
      exoticWeaponName: exoticWeaponName.present
          ? exoticWeaponName
          : Value(existing.exoticWeaponName),
      pinnedSuper:
          pinnedSuper.present ? pinnedSuper : Value(existing.pinnedSuper),
      softStatTargets: Value(
        encodeSoftStatTargetsJson(
          softStatTargets ?? existing.softStatTargets,
        ),
      ),
      updatedAt: Value(now),
    ),
  );

  if (tagIds != null) {
    await (db.delete(db.buildTags)..where((t) => t.buildId.equals(id))).go();
    for (final tagId in tagIds) {
      await db.into(db.buildTags).insert(
            BuildTagsCompanion.insert(buildId: id, tagId: tagId),
          );
    }
  }
  if (synergyTypes != null) {
    await (db.delete(db.buildSynergyTypes)
          ..where((t) => t.buildId.equals(id)))
        .go();
    await _insertSynergyTypes(db, id, synergyTypes, now);
  }

  return getBuild(db, userId, id);
}

/// Delete build (cascades variants/attachments/tags). Returns true if deleted.
Future<bool> deleteBuildRecord(
  AppDatabase db,
  int userId,
  String id,
) async {
  final n = await (db.delete(db.builds)
        ..where((t) => t.id.equals(id) & t.userId.equals(userId)))
      .go();
  return n > 0;
}
